atomisticparsers/h5md/parser.py

```diff
@@ -0,0 +1,787 @@
+    def __init__(self):
+        super().__init__(None)
+
+    @property
+    def filehdf5(self):
+        if self._file_handler is None:
+            try:
+                self._file_handler = h5py.File(self.mainfile, 'r')
+            except Exception:
+                self.logger.error('Error reading hdf5 file.')
+
+        return self._file_handler
+
+    def apply_unit(self, quantity, unit, unit_factor):
+
+        if quantity is None:
+            return
+        if unit:
+            unit = ureg(unit)
+            unit *= unit_factor
+            quantity *= unit
+
+        return quantity
+
+    def decode_hdf5_bytes(self, dataset, default):
+        if dataset is None:
+            return
+        elif type(dataset).__name__ == 'ndarray':
+            if dataset.size == 0:
+                return default
+            dataset = [val.decode("utf-8") for val in dataset] if isinstance(dataset[0], bytes) else dataset
+            dataset = [val.__bool__() for val in dataset] if type(dataset[0]).__name__ == 'bool_' else dataset
+        elif type(dataset).__name__ == 'bool_':
+            dataset = dataset.__bool__()
+        else:
+            dataset = dataset.decode("utf-8") if isinstance(dataset, bytes) else dataset
+        return dataset
+
+    def hdf5_attr_getter(self, source, attribute, path=None, default=None):
+        '''
+        Extracts attribute from object based on path, and returns default if not defined.
+        '''
+        if path:
+            section_segments = path.split('.')
+            for section in section_segments:
+                try:
+                    value = source.get(section)
+                    source = value[-1] if isinstance(value, list) else value
+                except Exception:
+                    return
+        value = source.attrs.get(attribute)
+        source = value[-1] if isinstance(value, list) else value
+        source = self.decode_hdf5_bytes(source, default) if source is not None else default
+
+        return source
+
+    def hdf5_getter(self, source, path, default=None):
+        '''
+        Extracts attribute from object based on path, and returns default if not defined.
+        '''
+        section_segments = path.split('.')
+        for section in section_segments:
+            try:
+                value = source.get(section)
+                unit = self.hdf5_attr_getter(source, 'unit', path=section)
+                unit_factor = self.hdf5_attr_getter(source, 'unit_factor', path=section, default=1.0)
+                source = value[-1] if isinstance(value, list) else value
+            except Exception:
+                return
+
+        if source is None:
+            source = default
+        elif type(source) == h5py.Dataset:
+            source = source[()]
+            source = self.apply_unit(source, unit, unit_factor)
+
+        return self.decode_hdf5_bytes(source, default)
+
+    def parse(self, quantity_key: str = None, **kwargs):
+        pass
```

Comment:
we can also interface the hdf5_getters to the get function here by populating parse something like:
```python
def parse(key, **kwargs):
    source = kwargs.get('source', self.filehdf5)
    path = kwargs.get('path')
    attribute = kwars.get('attribute')
    if attribute is not None:
        value = self.hdf5_attr_getter(source, path, attribute)
    else:
        value = self.hdf5_getter(source, path)
    # i do not know how to construct the full path here relative to root 
    full_path = f'...{path}.{attribute}'
    self._results[full_path] = value
```
This way, we also cache results.

Reply:
Is there an example that you can point me to that does something like this? All the parser classes that I can find either don't explicitly define parse() or pass within this function. I don't really understand this caching or what `full_path` is referring to

Reply:
you can have a look at the xml_parser.py in nomad.parsing.file_parser. The idea is to save the datasets in self._results[path.to.dataset] so that the next time you get(path.to.dataset), it is not fetched from the h5 file.